### ISSUE TITLE
how to load the features of smiles 

### ISSUE THREAD

**Issue by MiroWatson** [chemprop/chemprop#160]
I have noticed the steps of loading features of smiles in readme,but it will show  "ValueError: could not convert string to float: 'O1CC[C@@H](NC(=O)[C@@H](Cc2cc3cc(ccc3nc2N)-c2ccccc2C)C)CC1(C)C" when I try to run it.
this is my code :chemprop_train --data_path /home/user/test/chemprop-master/data/bace.csv --dataset_type classification --features_generator rdkit_2d_normalized --no_features_scaling --features_path /home/user/test/2/bace11.csv --save_dir /home/user/test/zwh/bace_checkpoints.
I want ti konw the correct way to load the features of smiles.

**Comment by cjmcgill** [comment]
I think I know what's causing that error. In your features file, you do not need a column with the smiles string and it will cause this error if you have one. Chemprop will assume that your features are listed in the same order as your smiles in the data input and associate the two that way. Because of this, there is no safety check that things are listed in the right order, so be careful that your features are in the right order and row positions.

**Comment by MiroWatson** [comment]
Thank you for your prompt reply.But after we changed the target file，a new error has occurred：**numpy.AxisError: axis 1 is out of bounds for array of dimension 1**，We thought it was a problem with the csv file at first, so we tried to change the csv file to an npy file, but the same error still appeared.This is the code we use: chemprop_train --data_path /home/user/test/chemprop-master/data/bace.csv --dataset_type classification --features_generator rdkit_2d_normalized --no_features_scaling --features_path /home/user/test/2/1.npy.If you can, can you please give me an example of how to express the characteristics, either csv file or npy file.Finally, thank you again for your prompt reply.

**Comment by cjmcgill** [comment]
A csv file should be able to work for your features, and is actually what I've always used. I will provide an example when I have the chance. But first, could you paste in the entire trace back for the error? That would be very helpful.

**Comment by MiroWatson** [comment]
I'm very glad that you can reply to me so quickly.
According to your statement, we changed back to the csv file,this is my complete error report.
`Traceback (most recent call last):
  File "/home/user/anaconda3/envs/chemprop/bin/chemprop_train", line 8, in <module>
    sys.exit(chemprop_train())
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/train/cross_validate.py", line 154, in chemprop_train
    cross_validate(args=TrainArgs().parse_args(), train_func=run_training)
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/utils.py", line 410, in wrap
    result = func(*args, **kwargs)
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/train/cross_validate.py", line 60, in cross_validate
    data = get_data(
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/data/utils.py", line 191, in get_data
    features_data = np.concatenate(features_data, axis=1)
  File "<__array_function__ internals>", line 5, in concatenate
numpy.AxisError: axis 1 is out of bounds for array of dimension 1
`
Please allow me to express my gratitude again, it is our honor to get your help.

**Comment by MiroWatson** [comment]
I'm very glad that you can reply to me so quickly.
According to your statement, we changed back to the csv file,this is my complete error report.

> Traceback (most recent call last):
  File "/home/user/anaconda3/envs/chemprop/bin/chemprop_train", line 8, in <module>
    sys.exit(chemprop_train())
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/train/cross_validate.py", line 154, in chemprop_train
    cross_validate(args=TrainArgs().parse_args(), train_func=run_training)
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/utils.py", line 410, in wrap
    result = func(*args, **kwargs)
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/train/cross_validate.py", line 60, in cross_validate
    data = get_data(
  File "/home/user/anaconda3/envs/chemprop/lib/python3.8/site-packages/chemprop/data/utils.py", line 191, in get_data
    features_data = np.concatenate(features_data, axis=1)
  File "<__array_function__ internals>", line 5, in concatenate
numpy.AxisError: axis 1 is out of bounds for array of dimension 1


Please allow me to express my gratitude again, it is our honor to get your help.

**Comment by MiroWatson** [comment]
We thought about it again. Maybe the title row of our target csv file is set incorrectly. The figure above is the title row we set. Do we need to set a few more columns later to store the features? If so, how many columns should be set?
![MC{2CB_JL1 _2T9_CSH154E](https://user-images.githubusercontent.com/75102374/114974645-6377d900-9eb5-11eb-85b1-ccbad28ddf78.png)

**Comment by cjmcgill** [comment]
You don't need to provide a blank file to store features from the feature generator. The argument --features_path points to a csv file where you supply additional features yourself. The feature generator and the features_path are two separate sources of features.

**Comment by hesther** [comment]
Also, if you want to save the generated features to a file, there is a script to do so in the scripts folder. I can send you an example if needed - other than that, if you don't actually want to read descriptors from a file, I agree with @cjmcgill, just try:

`chemprop_train --data_path /home/user/test/chemprop-master/data/bace.csv --dataset_type classification --features_generator rdkit_2d_normalized --no_features_scaling  --save_dir /home/user/test/zwh/bace_checkpoints`

**Comment by MiroWatson** [comment]
Thanks for your responding @hesther @cjmcgill.I have learned more from you. As you said, I want to save the generated features to a file, could you show me the example of that how to run the code.

**Comment by hesther** [comment]
Sure. Using the folder structure you mentioned, you would first run:

`python /home/user/test/chemprop-master/scripts/save_features.py --data_path /home/user/test/chemprop-master/data/bace.csv --save_path /home/user/test/2/bace11.npz --features_generator rdkit_2d_normalized`

which will precompute features specific to the provided input molecules, and saves them to a npz file. Now, you can run

`chemprop_train --data_path /home/user/test/chemprop-master/data/bace.csv --dataset_type classification --features_path /home/user/test/2/bace11.npz --no_features_scaling --save_dir /home/user/test/zwh/bace_checkpoints`

Let me know whether that works!

**Comment by TheZWH** [comment]
I saw this discussion and it intrigued me.I think this export feature generator is very useful. I also try to use the above process to test how to export rdkit_ 2D features.And I noticed that the combination of graph features and extra features was mentioned in this paper, so I want to know how to export graph features?I look forward to the pleasure of hearing from you.

**Comment by hesther** [comment]
Dear @EXAID-Parado, you can also export graph features. If you have a trained model, you can simply use `python fingerprint.py` (or `chemprop_fingerprint`) to load the model, run a bunch of molecules through the network and save the MPNN output (the graph features) to file. The Readme has some instructions on that (see [here](https://github.com/chemprop/chemprop#encode-fingerprint-latent-representation)) - let me know if you have additional questions!

**Comment by hesther** [comment]
Am closing this since I think the issue was resolved. Feel free to reopen if any more problems/questions arise!

**Comment by muammar** [comment]
> Am closing this since I think the issue was resolved. Feel free to reopen if any more problems/questions arise!

@hesther, I had a question -- are there plans to add support for ensembles? 

I am having this `ValueError`:

```
    raise ValueError("Fingerprint generation only supports one model, cannot use an ensemble")
ValueError: Fingerprint generation only supports one model, cannot use an ensemble

```

If not, is there a hacky way I can retrieve the representations? I really appreciate any help you can provide.

**Comment by hesther** [comment]
@muammar You can simply run `fingerprint.py` with each model separately by specifying the full path to the individual `model.pt` files, which will save the fingerprint for each model to a separate file. Let me know if you need further explanations!

**Comment by muammar** [comment]
> @muammar You can run `fingerprint.py` with each model separately by specifying the full path to the individual `model.pt` files, which will save the fingerprint for each model to a separate file. Let me know if you need further explanations!

That makes sense. For each model in the ensemble, there is a folder associated with it. Thanks!

**Comment by cjmcgill** [comment]
@muammar the ability to generate fingerprints with ensembles was implemented in #193. The error you quote should have been removed. Could you start a new issue report so we can troubleshoot it there?

**Comment by hesther** [comment]
Oh, I forgot we already pushed that, thanks @cjmcgill :)

**Comment by muammar** [comment]
> @muammar the ability to generate fingerprints with ensembles was implemented in #193. The error you quote should have been removed. Could you start a new issue report so we can troubleshoot it there?

I knew I had read there was an implementation for ensembles. Starting a new issue might not be necessary (? but just let me know if I should after I explain). I just updated my git repository to 491085fda1115f1edb53dd4e642b1ef65211adae, and instead of using `chemprop_fingerprint` from my  `pip` installation, I invoked `fingerprint. py`. Now it works. Thanks, @cjmcgill, and @hesther, for being so helpful :) 

PS. The version I currently have installed that did not work is the following: 

```
(py38) $ python3 -c "import chemprop; print(chemprop.__version__)"
WARNING:root:No normalization for BCUT2D_MWHI
WARNING:root:No normalization for BCUT2D_MWLOW
WARNING:root:No normalization for BCUT2D_CHGHI
WARNING:root:No normalization for BCUT2D_CHGLO
WARNING:root:No normalization for BCUT2D_LOGPHI
WARNING:root:No normalization for BCUT2D_LOGPLOW
WARNING:root:No normalization for BCUT2D_MRHI
WARNING:root:No normalization for BCUT2D_MRLOW
1.3.1
```

**Comment by cjmcgill** [comment]
@muammar , this is a new feature since the 1.3.1 release so I think this is expected behavior. Glad you're going to get use out of this new feature!
